### ISSUE TITLE
Bug in generic_multiedge_match

### ISSUE THREAD

**Issue by rstmr** [networkx/networkx#2114]
I ran into trouble when trying to check MultiDiGraphs for isomorphisms. The edges and nodes have  numerical and categorical attributes, so I used generic matches like this:

```
import nx.algorithms.isomorphism.matchhelpers as mh
em = mh.generic_multiedge_match(['id', 'flowMin', 'flowMax'], ['', 0, 0], [eq, close, close])
nm = mh.generic_node_match(['id', 'flowMin', 'flowMax'], ['', 0, 0], [eq, close, close])
return networkx.is_isomorphic(G1, G2, node_match=nm, edge_match=em)
```

Now, at some point I got an error, because `close` tried to compare tuples of the form (id, flowMin, flowMax). The responsible code comes from `networkx.isomorphisms.matchhelpers.generic_multiedge_match`:

```
def generic_multiedge_match(attr, default, op):
    """Returns a comparison function for a generic attribute.

    The value(s) of the attr(s) are compared using the specified
    operators. If all the attributes are equal, then the constructed
    function returns True. Potentially, the constructed edge_match
    function can be slow since it must verify that no isomorphism
    exists between the multiedges before it returns False.

    Parameters
    ----------
    attr : string | list
        The edge attribute to compare, or a list of node attributes
        to compare.
    default : value | list
        The default value for the edge attribute, or a list of
        default values for the dgeattributes.
    op : callable | list
        The operator to use when comparing attribute values, or a list
        of operators to use when comparing values for each attribute.

    Returns
    -------
    match : function
        The customized, generic `edge_match` function.

    Examples
    --------
    >>> from operator import eq
    >>> from networkx.algorithms.isomorphism.matchhelpers import close
    >>> from networkx.algorithms.isomorphism import generic_node_match
    >>> nm = generic_node_match('weight', 1.0, close)
    >>> nm = generic_node_match('color', 'red', eq)
    >>> nm = generic_node_match(['weight', 'color'],
    ...                         [1.0, 'red'],
    ...                         [close, eq])
    ...

    """

    # This is slow, but generic.
    # We must test every possible isomorphism between the edges.
    if nx.utils.is_string_like(attr):
        def match(datasets1, datasets2):
            values1 = [data.get(attr, default) for data in datasets1.values()]
            values2 = [data.get(attr, default) for data in datasets2.values()]
            for vals2 in permutations(values2):
                for xi, yi in zip(values1, vals2):
                    if not op(xi, yi):
                        # This is not an isomorphism, go to next permutation.
                        break
                else:
                    # Then we found an isomorphism.
                    return True
            else:
                # Then there are no isomorphisms between the multiedges.
                return False
    else:
        attrs = list(zip(attr, default)) # Python 3
        def match(datasets1, datasets2):
            values1 = []
            for data1 in datasets1.values():
                x = tuple( data1.get(attr, d) for attr, d in attrs )
                values1.append(x)
            values2 = []
            for data2 in datasets2.values():
                x = tuple( data2.get(attr, d) for attr, d in attrs )
                values2.append(x)
            for vals2 in permutations(values2):
                for xi, yi, operator in zip(values1, vals2, op):    <---------------
                    if not operator(xi, yi):
                        return False
            else:
                return True
    return match
```

I put an arrow on the line that's crucial. As far as I understand, values1 and vals 2 are lists of tuples of edge attributes (in my case lists of tuples of the form (id, flowMin, flowMax)). op is a list of comparison operators (in my case [eq, close, close]). Therefore, this breaks.
To my understanding, it should be something like this instead (starting from where I put the arrow):

```
# PLEASE SEE EDIT BELOW
for tuple1, tuple2 in zip(values1, vals2):
    for elem1, elem2, operator in zip(tuple1, tuple2, op):
        if not operator(elem1, elem2):
            return False
```

EDIT: Now, that I think about it from a "Why should it work?" and not "Why does it fail?" perspective: The suggested code above 'solves' the problem of incorrect comparison, but it does not produce the desired behavior of the function, because it would return False after having checked only one of all the possible isomorphisms between the multiedges.
Therefore, the loops starting from the line right above the arrow should be as follows:

```
            for vals2 in permutations(values2):
                for xi, yi in zip(values1, vals2):
                    if not all(map(lambda x,y,z: z(x,y), xi, yi, op)):
                        # This is not an isomorphism, go to next permutation.
                        break
                else:
                    # Then we found an isomorphism.
                    return True
            else:
                # Then there are no isomorphisms between the multiedges.
                return False
```

**Comment by chebee7i** [comment]
@rstmr nice find.  You might be the first person to use this particular feature, especially for multidigraphs, ...in many years.  Would you be able to submit a pull request with this fix implemented, along with a unit test? No problem, if that is not a possibility. Just let us know.

**Comment by jfinkels** [comment]
Love the nested for/else.

**Comment by rstmr** [comment]
@chebee7i Yes, I'll take care of that later.
@jfinkels I got that from how it's handled for a single attribute right above. For/else allows nice code, but I often forget that, too.

**Comment by rstmr** [comment]
I added a pull request: https://github.com/networkx/networkx/pull/2124
Fix is as above. But, can anyone explain to me why the if/else is implemented with `if nx.utils.is_string_like(attr)`?

**Comment by chebee7i** [comment]
This was just to simplify the use case where `attr='weight'`; that is, a single attr instead of a list. In hindsight, we probably should just do something like:

```
if nx.utils.is_string_like(attr):
    attr = [attr]
    value = [value]
    op = [op]
```

If that makes sense to you and you agree, let's do that in #2124 

and then proceed with only one implementation.

**Comment by rstmr** [comment]
I added that to the commit.
My question regarding `nx.utils.is_string_like(attr)` was more directed towards: Why did you implement this custom function that checks if something is 'string-like' instead of using some Pytho build-in type check to distinguish between strings and lists?

**Comment by chebee7i** [comment]
The main reason was because that is how NetworkX preferred to test for a string. 

However, the rationale for the particular implementation of `is_string_like` has more to do with performing a test that does not rely explicitly on type-checking.  See:
 https://www.safaribooksonline.com/library/view/python-cookbook-2nd/0596007973/ch01s04.html 
for a more thorough description---in particular, take note of the issues with detecting when `attr` is a `UserString`. Also, keep in mind that this `is_string_like` function was written many, many years ago at a time when the Python world wasn't as clean/consistent as it is now. :)

**Comment by jfinkels** [comment]
Fixed in pull request #2114.
